Improved onError handling

dlmr opened the pull request:
This will prevent the application to get stuck in a loading state and requiring manual interaction.

@PAkerstrand Does this seem like a good default for you?

dlmr committed:
Will now always abort when an error happens

PAkerstrand commented on src/RedialContext.js:
```diff
@@ -188,7 +192,10 @@ export default class RedialContext extends Component {
+      .catch((err) => {
+        this.props.onError(err, bail() || 'other');
+        this.abort(true);
+      });
```
While I agree that we should abort (since we're no longer loading), I'm wondering whether we should also transition back to the previous location. If we abort, then we'll `setState`, which will cause a `render()`, which will render the newly matched location, but with data that failed to fetch. Perhaps that is not our concern, or maybe it is? Either way, I'd say it warrants further discussion

dlmr replied:
Yes, that would probably be good! I'm all for this change!

Maybe it should be an option, since a scenario might be that the user reloads the page and would then get the correct resource?

PAkerstrand replied:
Can you make an example where for one route, the blocking fetch fails, and for another route the deferred fetch fails? I think we might want to handle these differently...

dlmr committed:
Updated the example servers to log the error that might happen

dlmr committed:
Now runs defer and blocking separately in parallel and better onError

- Runs defer and blocking separately in parallel, before they where part of the same promise chain meaning if one failed the other would be discarded
- Better onError the power to reload the page for example on error or go back
- onCompleted now gets a arguments that will be either blocking or deferred depending on what completed

dlmr committed:
Updated simple example with new paths that fail on the client

This to test the new onError system mainly.

dlmr commented:
Made the error handling more generic where users can implement it as they see fit for their use case.

**Example**

``` js
const forcePageReloadOnError = true;
const goBackOnError = false;

// Function that can be used as a setting for useRedial
function onError(err, { abort, blocking, reason, router }) {
  if (process.env.NODE_ENV !== 'production') {
    console.error(reason, err);
  }

  // We only what to do this if it was a blocking hook that failed
  if (blocking) {
    if (forcePageReloadOnError) {
      window.location.reload();
    } else if (goBackOnError) {
      router.goBack();
    }
    // Abort current loading automatically
    abort();
  }
}
```

dlmr committed:
Improved the example for onError to only take action on certain reasons

dlmr committed:
Renamed RedialContextContainer to RedialContainer and added displayName

By using displayName the component will keep its name even after minification

dlmr committed:
Added onCompleted when both blocking and deferred has completed, "all"

dlmr committed:
Changed so onCompleted for blocking will always be invoked before defer

dlmr committed:
We now only run defer onError if blocking completed successfully

This makes parallel and non parallel behaviour consistent

dlmr committed:
Made sure that aborting from an error only cancels the corresponding req

This solves an issue where aborting an then quickly making a new request, resulting in that both the old and the new request was aborted

dlmr committed:
Renamed blocking => beforeTransition and defer => afterTransition

Closes #10

dlmr commented:
The question is if we should change `onCompleted(type)` to `beforeTransitionDidComplete()` and `afterTransitionDidComplete()`. Thoughts @PAkerstrand?

PAkerstrand commented on src/RedialContext.js:
```diff
@@ -17,6 +17,8 @@ function hydrate(renderProps) {
 }
 
 export default class RedialContext extends Component {
+  static displayName = 'RedialContext';
```
Why do we need this? In `__DEV__` it'll be inferred automatically, and in `production` we don't want/need the name to not be minimized?

dlmr replied:
Correct, we would like to keep the name even in a minified build.

PAkerstrand replied:
ok

PAkerstrand commented on src/RedialContext.js:
```diff
@@ -165,36 +174,60 @@ export default class RedialContext extends Component {
+      .then(() => {
+        if (this.state.beforeTransitionCompleted) {
+          this.props.onCompleted('afterTransition');
+        }
```
If we are in `parallel`-mode, won't we have a risk of calling `onCompleted` twice with `'afterTransition'`? Here, and line L280

dlmr replied:
True, this could probably happen currently depending how how the promises are invoked. Need to look closer at this.

PAkerstrand commented on src/RedialContainer.js:
```diff
@@ -1,14 +1,16 @@
 
-export default class RedialContextContainer extends React.Component {
+export default class RedialContainer extends Component {
+  static displayName = 'RedialContainer';
```
Needed?

dlmr replied:
If we want the component to not be called something like `t` in React devtools.

PAkerstrand replied:
ok

PAkerstrand reviewed (commented):
Looks good overall! 👍

dlmr committed:
Made sure we have consistent behaviour when running in parallel and not

Made sure that we only run onCompleted for afterTransition once
Made sure we wait with showing error for afterTransition until beforeTransition have succeded
